### DISCUSSION TITLE
When using free drawing, the brush color has transparency. How can I cover the previous graphics?

### DISCUSSION THREAD

**Discussion by ysl-ccc** [fabricjs/fabric.js#9688] [Q&A] [answered]
When using free drawing, the brush color has transparency. How can I cover the previous graphics?

With free drawing, the pen color has transparency, and when overlapping graphics are used, the color being drawn is dominant.

Similar to the drawing form when the brush has no transparency, the shape currently being drawn is displayed each time, and the remaining shapes will be covered in the overlapping area.

**Comment by ShaMan123** [comment]
Please provide a live repro and visuals

**Reply by ysl-ccc** [reply]
https://codesandbox.io/p/sandbox/fabric-drawing-fmfl5p

![test](https://github.com/fabricjs/fabric.js/assets/63445570/1131d740-7658-4eb6-abb2-a1b99fb3ea2c)

The color of the brush is transparent. At the intersection of two graphics, I only want to display the color of one of them. How to achieve this?

For example: it is displayed like this:
![test1](https://github.com/fabricjs/fabric.js/assets/63445570/ef030c8a-35f9-4f34-bcd7-10255592f7b7)

**Comment by ShaMan123** [comment, marked as answer]
Take a look at [this](https://codesandbox.io/p/sandbox/fabric-drawing-forked-kplngf?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clt2kj1vu00063b6iau186p8u%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clt2kj1vt00023b6itd3z7eav%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clt2kj1vt00033b6ikwv7zfss%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clt2kj1vu00053b6if0k9679i%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clt2kj1vt00023b6itd3z7eav%2522%253A%257B%2522id%2522%253A%2522clt2kj1vt00023b6itd3z7eav%2522%252C%2522tabs%2522%253A%255B%255D%257D%252C%2522clt2kj1vu00053b6if0k9679i%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clt2kj1vt00043b6itk04awqi%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clt2kj1vu00053b6if0k9679i%2522%252C%2522activeTabId%2522%253A%2522clt2kj1vt00043b6itk04awqi%2522%257D%252C%2522clt2kj1vt00033b6ikwv7zfss%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clt2kj1vt00033b6ikwv7zfss%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

I used globalCompositeOperation to achieve what you want
A gift from me
In case you will draw other objects and not only paths makes sure to group all paths together so that the globalCompositeOperation affects only them and turn objectCaching to true on the group for context isolation

**Comment by ysl-ccc** [comment]
I need to maintain the transparency of the graphics generated by the brush, but this solution cannot be achieved

**Reply by ShaMan123** [reply]
In the demo I shared I believe it works, doesn't it?

**Comment by ysl-ccc** [comment]
Your method only works when there is no background. When I add a background, it doesn't work. You can see my [example](https://codesandbox.io/p/sandbox/fabric-drawing-forked-cjlrj2).

The solution I thought of is this: after the drawing is completed, merge the paths of objects of the same color so that their overlapping areas will not have different colors.

**Reply by ShaMan123** [reply]
I can't access your example, did you share it?
Your solution is similar to what I suggested but better in fact
> In case you will draw other objects and not only paths makes sure to group all paths together so that the globalCompositeOperation affects only them and turn objectCaching to true on the group for context isolation

**Reply by ShaMan123** [reply]
This uses a different approach for the xor operation:

[![Edit fabric-drawing (forked)](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/p/sandbox/fabric-drawing-forked-37j4x4)

**Reply by ysl-ccc** [reply]
I just added this line of code:

`
    fabric.Image.fromURL(bg, (img) => {
      canvas.setBackgroundImage(img, canvas.renderAll.bind(canvas));
    });
    canvas.freeDrawingBrush = new XorBrush(canvas);
`

![3](https://github.com/fabricjs/fabric.js/assets/63445570/79e5eb04-4674-4447-abf5-6cca79ae56f3)

The left side is the code for adding the background image, and the right side is the graphics for drawing two strokes. The color display of the overlapping area of the two graphics is still incorrect.

But the current effect still superimposes the colors in the overlapping areas of the two graphics, instead of only displaying the color of one of them.

**Reply by ShaMan123** [reply]
So it seems the xor operation is not what you need
You can try to use a remove color filter perhaps, depends on your app's needs, or else maybe it is enough to hide the background color while drawing

**Reply by ShaMan123** [reply]
Maybe you should use the eraser brush approach
You erase the path using `destination-out` and then draw it with a pattern of what you want to see

**Reply by ysl-ccc** [reply]
Similar problems occur when there is a background or a pen color with transparency. My solution is to splice the paths of objects of the same color together to form one object after each brush drawing is completed, so as to avoid the display problem of overlapping areas.
